### DISCUSSION TITLE
Liveprovider subscription update on filter param changed

### DISCUSSION THREAD

**Discussion by quick691fr** [refinedev/refine#6366] [General]
Hi guys,

I'm trying to make a useList working with liveProvider and supabase when filter changed.

I have this code retrieving widgets from supabase :
```
const { data: widgetsData } = useList<IWidget>({
    resource: "widgets",
    filters: [
      {
        field: "dashboard_id",
        operator: "eq",
        value: dashboardId,
      },
    ],
  });
```

dashboardId is a prop of the component containing this useList, so basically, I can request this component with different dashboardId.

I checked the websocket and I got a subscription for realtime events :
```
realtime:resources/widgets:*:dashboard_id=eq.2
```

and realtime changes on widgets of dashboard 2 is working fine.

But when I changed my dashboardId Props by selecting a new dashboard on an external Select, the realtime subscription is not updated and I didn't get LiveEvents on new dashboard. My guess is the realtime subscription should reflect the filter change ??

Thanks

**Comment by alicanerdurmaz** [comment]
Hello @quick691fr,

There is a similar issue [here](https://github.com/refinedev/refine/issues/6360).

Removing filters from the subscription will be a solution. This way, the `widgets` table will be invalidated on any change, regardless of the filters.

This might cause unnecessary invalidation in some cases, but I don't think it's a big problem.

**Reply by quick691fr** [reply]
Hi @alicanerdurmaz,

That's what I was wondering, removing filter from subscription, but then I was thinking with like 10000 users on the dashboard, it's gonna trigger refresh the 10 000 views instead of a single dashboard. Any clue or fix going on ?

**Reply by alicanerdurmaz** [reply]
After we remove `filters` from the subscription query, Refine will invalidate all queries with the `users` key regardless of filters and refetch `users` using the latest filters. 

What kind of dashboard has 10.000 users? 🤔 Probably you need an infinite list or paginated table to show that. In that case, the app will fetch like 10 or 25 user.

**Reply by quick691fr** [reply]
No dashboard has 10000 users.

I have few dashboards per user.
When a user is going to the dashboard part of the webapp, it displays a unique dashboard let say id:1.

If I have 10000 users connected on my website displaying their own dashboard (which get different ids) but I remove the filter from subscription, it means any update on a widget of a single user dashboard, will trigger invalidation on all users dashboards so 10000 refresh instead of 1.

**Reply by alicanerdurmaz** [reply]
Thank you for your detailed explanation. The subscription filters must work in this scenario! Could you please provide a simple reproducible example? We'd like to debug it using your use case.

**Reply by quick691fr** [reply]
Hi @alicanerdurmaz, here is an example of failling liveprovider : https://github.com/quick691fr/share-refine-liveprovider 

You have to run `npm install` and  `npm run dev`.

Then on the website, sign in with embedded credentials and go to liveproviders.
You have a Select on top to select the category of posts, default selected is category 1 "Laptops and Desktopsss" you can check in inspector the websocket is subscribing to live updates: ` "topic": "realtime:resources/blog_posts:*:categoryId=eq.1",`. If you update title of a post in this category, live update is triggered (since in correct category) and title list is updated automatically as expected.

But if you change category in the select and choose category 15 aka "Automotive and Industrial", the websocket is not subscribing to live updates: ` "topic": "realtime:resources/blog_posts:*:categoryId=eq.15",` and then all updates done on a post of this category don't trigger liveupdate.

Hope it helps you to fix it.

**Reply by quick691fr** [reply]
ping ?

**Reply by alicanerdurmaz** [reply]
Hello @quick691fr, Thanks for the detailed explanation.

Actually, the live provider is working as expected. I can verify this through Chrome's network tab but we have some issues with Ant Design's `<Select />`, It throws this error:
```md
hook.js:591 Warning: `label` of `value` is not same as `label` in Select options. Error Component Stack
```

I'm still looking for a solution, I will inform you when I solve it.

Working example with native HTML select:
```tsx
'use client'

import { PostList } from '@components/post-list'
import { useSelect } from '@refinedev/antd'
import type { BaseOption } from '@refinedev/core'
import { Flex, Select } from 'antd'
import { useState } from 'react'

export default function LiveProviderList() {
  const [categoryId, setCategoryId] = useState<number>(1)
  const { selectProps } = useSelect({
    resource: 'categories',
    defaultValue: categoryId,
  })

  return (
    <Flex vertical justify='center' align='center' style={{ width: '100%' }}>
      <select
        onChange={(e) => {
          setCategoryId(Number(e.target.value))
        }}>
        {selectProps.options?.map((option) => (
          <option key={option.value} value={option.value as string}>
            {option.label}
          </option>
        ))}
      </select>
      <PostList categoryId={categoryId} />
    </Flex>
  )
}
```

**Reply by alicanerdurmaz** [reply]
I've found a workaround to fix Ant Design's `<Select />`. When I don't render components on the loading state it's working but this isn't an ideal solution. 

It looks like a `state` problem, I will look into it.


```tsx
'use client'

import { PostList } from '@components/post-list'
import { useSelect } from '@refinedev/antd'
import type { BaseOption } from '@refinedev/core'
import { Flex, Select, Skeleton } from 'antd'
import { useState } from 'react'

export default function LiveProviderList() {
  const [categoryId, setCategoryId] = useState<number>(5)
  const { selectProps, query, defaultValueQuery } = useSelect({
    resource: 'categories',
    defaultValue: categoryId,
  })

  const loading = query.isFetching || defaultValueQuery.isFetching

  return (
    <Flex vertical justify='center' align='center' style={{ width: '100%' }}>
      {!loading ? (
        <>
          <Select
            style={{ width: '250px' }}
            defaultValue={categoryId as unknown as BaseOption}
            {...selectProps}
            onSelect={(value) => {
              setCategoryId(Number(value))
            }}
          />
          <PostList categoryId={categoryId} />
        </>
      ) : (
        <Skeleton.Input style={{ width: '250px' }} active />
      )}
    </Flex>
  )
}

```
![image](https://github.com/user-attachments/assets/d7e7f6f4-344f-45d8-ae8d-42d2e8743920)

**Reply by quick691fr** [reply]
Strange, I copy paste your code and still have no realtime topic changed, what version are you using ?
![2024-10-15 111556](https://github.com/user-attachments/assets/879563ef-afb5-4065-ac94-565aca6c6d56)

**Reply by quick691fr** [reply]
My bad, apparently, the one doing the job is :

const { selectProps, query, defaultValueQuery } = useSelect({
    resource: "categories",
    **_defaultValue: categoryId,_**
  });

**Reply by quick691fr** [reply]
Strange that the defaultValue on category useSelect fix realtime subscription on blog_posts useList in another component ??

**Reply by alicanerdurmaz** [reply]
Oh, no. Sorry for the confusion.

We have 2 issues in this case:
1.  Ant Design's `<Select />` gives an error when `options` are changed. Some other people [reported](https://github.com/ant-design/ant-design/issues/40205) that too.
2. `<PostList categoryId={categoryId} />` uses `useList` with a resource of `blog_posts` and filters query with `categorId` but `useList` doesn't subscribe again when filters are changed. This is a bug from Refine, re-mounting the component is working but is not an ideal solution.

Changing `defaultValue` is working because it updates `defaultValueQuery.isFetching` to `true` and we are re-mounting the `<PostList />`. It's a bit tricky and confusing workaround but I believe it's good while we work on this bug.

**Reply by alicanerdurmaz** [reply]
This works too. I think this is a more descriptive example.


Example with unique `key` prop:
```tsx
'use client'

import { PostList } from '@components/post-list'
import { useSelect } from '@refinedev/antd'
import type { BaseOption } from '@refinedev/core'
import { Flex, Select } from 'antd'
import { useState } from 'react'

export default function LiveProviderList() {
  const [categoryId, setCategoryId] = useState<number>(5)
  const { selectProps } = useSelect({
    resource: 'categories',
  })

  return (
    <Flex vertical justify='center' align='center' style={{ width: '100%' }}>
      <Select
        key={`select-${categoryId}`}
        style={{ width: '250px' }}
        defaultValue={categoryId as unknown as BaseOption}
        {...selectProps}
        onSelect={(value) => {
          setCategoryId(Number(value))
        }}
      />
      <PostList key={`post-list-${categoryId}`} categoryId={categoryId} />
    </Flex>
  )
}
``

**Comment by amerryma** [comment]
I too am running into this issue. Switching to useList _does_ work because it listens for all changes on my table.

Desired:
```typescript
  const { data: job } = useOne<JobResult, HttpError, JobResult>({
    resource: 'job_results',
    liveMode: 'auto',
    id: jobKey,
  })
```

Workaround:
```typescript
  const { data: jobs } = useList<JobResult, HttpError, JobResult>({
    resource: 'job_results',
    liveMode: 'auto',
    pagination: {
      pageSize: 1,
    },
    sorters: [{ field: 'created_at', order: 'desc' }],
  })
  console.log('jobs', jobs)
  const job = jobs?.data?.find((job) => job.id === jobKey)
```

**Reply by alicanerdurmaz** [reply]
Hello @amerryma,

Can you try this?
```tsx
const { data: job } = useOne<JobResult, HttpError, JobResult>({
  resource: 'job_results',
  liveMode: 'auto',
  id: jobKey,
  liveParams: {
    filters: [
      {
        field: 'id',
        operator: 'eq',
        value: jobKey,
      },
    ],
  },
})
```

Sorry for your trouble, Refine should add an ID filter automatically. We will look into this.

**Reply by quick691fr** [reply]
> I too am running into this issue. Switching to useList _does_ work because it listens for all changes on my table.
> 
> Desired:
> 
> ```ts
>   const { data: job } = useOne<JobResult, HttpError, JobResult>({
>     resource: 'job_results',
>     liveMode: 'auto',
>     id: jobKey,
>   })
> ```
> 
> Workaround:
> 
> ```ts
>   const { data: jobs } = useList<JobResult, HttpError, JobResult>({
>     resource: 'job_results',
>     liveMode: 'auto',
>     pagination: {
>       pageSize: 1,
>     },
>     sorters: [{ field: 'created_at', order: 'desc' }],
>   })
>   console.log('jobs', jobs)
>   const job = jobs?.data?.find((job) => job.id === jobKey)
> ```

Yes, it works without filter but then you subscribe to whole * updates which is not efficient regarding the database, specially if you have frequent updates.

**Reply by amerryma** [reply]
Ya, definitely not ideal. And to be clear, my failure doesn't happen as long as the ID never changes while the component is mounted, but once the ID changes, it fails to switch to the new subscription.

I will try the `liveParams` soon.

**Reply by amerryma** [reply]
Ya `liveParams` still doesn't change the subscription if `jobKey` were to change without remounting the component.

**Reply by alicanerdurmaz** [reply]
Thanks for the detailed explanation. We will look into this remounting issue.

**Reply by quick691fr** [reply]
> Thanks for the detailed explanation. We will look into this remounting issue.

@alicanerdurmaz any news or issue #?

**Reply by quick691fr** [reply]
ping ?

**Reply by alicanerdurmaz** [reply]
Hello, there are a couple of things that need to be fixed regarding this issue. For example, a PR has already been opened for it: https://github.com/refinedev/refine/pull/6573. However, things have been quite busy lately, and I haven’t found the time to address this problem. 

We welcome all PRs, and if you'd like to work on this, we'd be more than happy. 🙌
